When requests fails (while using timeout addition), preserve original stacktrace

esse opened the pull request:
Otherwise the stacktrace points to;

```
(tesla) lib/tesla/middleware/timeout.ex:54: Tesla.Middleware.Timeout.repass_error/1
(tesla) lib/tesla/middleware/timeout.ex:32: Tesla.Middleware.Timeout.call/3
(...)
```

which isn't very helpful.

teamon commented:
👋  Thank you for your contribution!

I'm all for improving the stacktrace, as long as we do **not** hide an important trace.
We would definitely need tests for this.

Btw, `__STACKTRACE__` is available only in elixir 1.7+ while tesla supports `"~> 1.5"`.

esse commented:
@teamon thanks for your feedback!

I've changed `__STACKTRACE__` to `System.stacktrace/0` for compatibility and I've added tests for expected behaviour.

esse commented:
@teamon can you look into this? It would be very helpful for us to have this merged, since for now we're using forked version.

teamon commented:
Please rebase on master and ensure tests are passing :)

esse commented:
@teamon - thanks, done 👍
